airbyte-integrations/connectors/source-posthog/source_posthog/streams.py

```diff
@@ -0,0 +1,306 @@
+
+        stream_state = stream_state or {}
+        params = super().request_params(stream_state=stream_state, **kwargs)
+        params["order"] = f"-{self.cursor_field}"
+        return params
+
+
+class Cohorts(IncrementalPosthogStream):
+    """
+    normal ASC sorting. But without filters like `since`
+    So we need to query all anyway, but we may skip already written reords for incremental.
+    """
+
+    cursor_field = "id"
+
+    def path(self, **kwargs) -> str:
+        return "cohort"
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        resp_json = response.json()
+        if "next" in resp_json and resp_json["next"]:
+            next_query_string = urllib.parse.urlsplit(resp_json["next"]).query
+            params = dict(urllib.parse.parse_qsl(next_query_string))
+            if params:
+                return params
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
+        """
+        Return the latest state by comparing the cursor value in the latest record with the stream's most recent state object
+        and returning an updated state object.
+        """
+        current_stream_state = current_stream_state or {}
+        latest_state = latest_record.get(self.cursor_field)
+        current_state = current_stream_state.get(self.cursor_field) or latest_state
+        return {self.cursor_field: max(latest_state, current_state)}
+
+    def parse_response(self, response: requests.Response, stream_state: Mapping[str, Any], **kwargs) -> Iterable[Mapping]:
+        response_json = response.json()
+        data = response_json.get(self.data_field)
+        if data and stream_state:
+            last_record_curfield = data[-1][self.cursor_field]
+            last_stream_curfield = stream_state[self.cursor_field]
+            if last_stream_curfield >= last_record_curfield:
+                yield from []
+            else:
+                first_record_curfield = data[0][self.cursor_field]
+                if first_record_curfield > last_stream_curfield:
+                    yield from data
+                else:
+                    slice_ = [i for i in data if i[self.cursor_field] > last_stream_curfield]
+                    yield from slice_
+
+        else:
+            yield from data
+
+
+class Elements(PosthogStream):
+    data_field = "elements"
+
+    def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
+        return "element/stats"
+
+    def parse_response(self, response: requests.Response, stream_state: Mapping[str, Any], **kwargs) -> Iterable[Mapping]:
+        response_json = response.json()
+        elements = [i[self.data_field][0] for i in response_json]
+        yield from elements
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        """
```

Comment:
We have 4 different overwrites of this method. Can we group the conditions and override once in the `PosthogStream` class?

Reply:
actually this is dependent on endpoint behaviour. 
Ways of implementations from Posthog:
1) it may return link to next result_set in `next` attribute of response`
2) it may send pagination parameters as object in `pagination` attribute
3) in some case pagination compeltely inherits parent logic, but need to be interrupted on condition (reverse incremental).
4) it may be absent in all

Make some small changes in this.